lang_qc/db/helper/well.py

```diff
@@ -110,94 +110,114 @@ def exists(self) -> bool:
+
+        return self.qc_types[qc_type_name]
 
     @cached_property
-    def product_definition(self) -> Dict:
+    def qc_states(self) -> Dict:
         """
-        A cached representation of the well as a dictionary with two keys,
-        `id` - a unique product id and `json` - a json string representing
-        this well, see `npg_id_generation.PacBioEntity` for details. This
-        property is an outcome of a ligh-weight computation that does not
-        involved querying the database. The property is computed lazily.
+        A cached dictionary of QC state dictionary rows, where the QcStateDict objects
+        representing the database rows are values abd the string descriptions of the
+        state are keys.
         """
 
-        pbe = PacBioEntity(run_name=self.run_name, well_label=self.well_label)
-        return {"id": pbe.hash_product_id(), "json": pbe.json()}
+        db_states = self.session.execute(select(QcStateDict)).scalars().all()
+        return {row.state: row for row in db_states}
 
-    def valid_qc_state_dict(self, qc_state: str) -> QcStateDict:
+    def qc_state_dict_row(self, qc_state_name: str) -> QcStateDict:
         """
         Given a description of the QC state, returns an object that
         corresponds to a relevant row in the `qc_state_dict` table.
         Raises a `ValidationError` if no corresponding dictionary record
         is found.
         """
 
-        valid = (
-            self.session.execute(
-                select(QcStateDict).where(QcStateDict.state == qc_state)
-            )
-            .scalars()
-            .one_or_none()
-        )
+        if qc_state_name not in self.qc_states.keys():
+            raise InvalidDictValueError(f"QC state '{qc_state_name}' is invalid")
 
-        if valid is None:
-            raise InvalidDictValueError(f"QC state '{qc_state}' is invalid")
+        return self.qc_states[qc_state_name]
 
-        return valid
 
-    def valid_qc_type(self, qc_type: str) -> QcType:
+class WellQc(QcDictDB):
+    """
+    A data access class for routine SQLAlchemy operations on PacBio well
+    QC data in LangQC database.
+    """
+
+    run_name: str = Field(title="Name of the run as known in LIMS")
+    well_label: str = Field(title="Well label as known in LIMS and SmrtLink")
+
+    class Config:
+        allow_mutation = False
+        # A workaround for a bug in pydantic in order to use the cached_property
+        # decorator.
+        keep_untouched = (cached_property,)
+
+    @cached_property
+    def seq_product(self) -> SeqProduct:
         """
-        Given a description of the QC type, returns an object that
-        corresponds to a relevant row in the `qc_type` table.
-        Raises a `ValidationError` if no corresponding record is found.
+        A cached instance of the lang_qc.db.qc_schema.SeqProduct object.
+        If the corresponding database record exists, this object corresponds
```

Comment:
This is not an object, it's a property. Therefore more simply "Returns a pre-existing SeqProduct, or creates a new one."?